### PR TITLE
TE_494-Creation_of_incorrect_Import_statements

### PR DESCRIPTION
**Pull request by georgportwich**

### CODE REVIEW

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend [hunk 1]
```diff
@@ -182,7 +185,8 @@ class TclModelDragAndDropUpdater {
 	 * the modification of the import section must take place before actually adding the new
 	 * test step component.
 	 */
-	public def void updateImports(TclModel tclModel, Component droppedObject, String qualifiedName) {
+	public def void updateImports(TclModel tclModel, EObject dropTarget, Component droppedObject,
+		String qualifiedName) {
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: I can't see to which object the `qualifiedName` belongs. Why do we need to pass this as a parameter anyway? We can use the qualifiedNameProvider to retrieve it.

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend [hunk 2]
```diff
@@ -195,74 +199,88 @@ class TclModelDragAndDropUpdater {
 		val droppedAlreadyImportedViaWildcard = tclModel.wildcardImports.exists [
 			resolveType(tclModel, simpleName)?.qualifiedName == qualifiedName
 		]
+		val testStepContext = dropUtils.searchTargetTestStepContext(tclModel, dropTarget)
+		var droppedAlreadyImportedByCurrentComponent = false
+		var droppedAlreadyImportedByTestCase = false
+		if (testStepContext instanceof ComponentTestStepContext) {
+			droppedAlreadyImportedByCurrentComponent = qualifiedName.equals(
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: Can we extract those decisions to separate methods?

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend [hunk 3]
```diff
@@ -195,74 +199,88 @@ class TclModelDragAndDropUpdater {
...
 		val importWanted = clashingImportedTypes.nullOrEmpty && !wildcardRemovalTookPlace &&
-			!droppedAlreadyImportedViaWildcard
+			!droppedAlreadyImportedViaWildcard && !droppedAlreadyImportedByCurrentComponent &&
+			!droppedAlreadyImportedByTestCase
 
 		if (importWanted) {
-			val added = importSection.addImport(qualifiedName)
-			if (!added) { // e.g. this import already exists
-				return // no further rewrite of import section necessary
+				val added = importSection.addImport(qualifiedName)
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: wrong indentation?

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend [hunk 4]
```diff
@@ -195,74 +199,88 @@ class TclModelDragAndDropUpdater {
...
-	 */
-	private def void saveUpdate(RewritableImportSection importSection, TclModel tclModel) {
-		if (tclModel.importSection === null) {
-			// importSection.update does not automatically create the import section itself
-			tclModel.importSection = xtypeFactory.createXImportSection
+		/**
+		 * make a safe update of the import section of this tcl model and ensure that 
+		 * the tclModel.importSection is null, if no imports remain 
+		 */
+		private def void saveUpdate(RewritableImportSection importSection, TclModel tclModel) {
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: Was there a formatter bug here?

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend [hunk 5]
```diff
@@ -182,7 +185,8 @@ class TclModelDragAndDropUpdater {
 	 * the modification of the import section must take place before actually adding the new
 	 * test step component.
 	 */
-	public def void updateImports(TclModel tclModel, Component droppedObject, String qualifiedName) {
+	public def void updateImports(TclModel tclModel, EObject dropTarget, Component droppedObject,
```
Comment (gunther-bachmann) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: why is the dropTarget needed? dropping A on the first or the second context should not make a difference in respect to the import statements, right?
component A:
- do this
- do that

component full.qual.B
- do some other

Comment (georgportwich) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: Ticket updated

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend [hunk 6]
```diff
@@ -195,74 +199,88 @@ class TclModelDragAndDropUpdater {
...
 		]
+		val testStepContext = dropUtils.searchTargetTestStepContext(tclModel, dropTarget)
+		var droppedAlreadyImportedByCurrentComponent = false
+		var droppedAlreadyImportedByTestCase = false
+		if (testStepContext instanceof ComponentTestStepContext) {
+			droppedAlreadyImportedByCurrentComponent = qualifiedName.equals(
+				qualifiedNameProvider.getFullyQualifiedName(testStepContext.component).toString())
+			droppedAlreadyImportedByTestCase = qualifiedName.equals(
+				qualifiedNameProvider.getFullyQualifiedName(testStepContext.eContainer.eContainer.eContainer).toString())
+		}
```
Comment (georgportwich) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: Exception included

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend (hunk 4 above)
Comment (georgportwich) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: done

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend (hunk 3 above)
Comment (georgportwich) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: yes - fixed

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend (hunk 1 above)
Comment (georgportwich) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: Renamed to qualifiedNameOfDroppedObject

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend (hunk 2 above)
Comment (georgportwich) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/TclModelDragAndDropUpdater.xtend]: fixed

**Pull request merged**